File: 200_Number_of_Islands/200.py
class Solution:
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        import sys
        if sys.version[0] == "2":
            from Queue import Queue
        else:
            from queue import Queue
        Q = Queue()
        engaged = set()
        row = len(grid)
        if row == 0: 
            return 0
        col = len(grid[0])
        if col == 0:
            return 0
        ans = 0
        for i in range(row):
            for j in range(col):
                # this earth has not been reached
                if (i, j) not in engaged and grid[i][j] == "1":
                    ans += 1
                    Q.put((i, j))
                    engaged.add((i, j))
                    while not Q.empty():
                        x, y = Q.get()
                        for x_, y_ in zip([1,0,-1,0], [0,1,0,-1]):
                            xx = x + x_
                            yy = y + y_
                            if 0<= xx < row and 0 <= yy < col and (xx, yy) not in engaged and grid[xx][yy] == "1":
                                Q.put((xx, yy))
                                engaged.add((xx, yy))
        return ans

File: 200_Number_of_Islands/200.cpp
//std没有对unordered_set实现hash函数
//需要自己实现并放入unordered_set的template
struct pairhash {
public:
  template <typename T, typename U>
  std::size_t operator()(const std::pair<T, U> &x) const
  {
    return std::hash<T>()(x.first) * 31 + std::hash<U>()(x.second);
  }
};
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        //walk all the connected earth
        queue<pair<int, int>> Q;
        //the earth that has been searched
        unordered_set<pair<int, int>, pairhash> used;
        //direction matrix
        vector<int> x_ = {1,0,-1,0};
        vector<int> y_ = {0,1,0,-1};
        //special case
        int row = grid.size();
        if(row == 0)
            return 0;
        int col = grid[0].size();
        if(col == 0)
            return 0;
        int ans = 0;
        //core
        for(int i = 0; i < row; ++i){
            for(int j = 0; j <col; ++j){
                auto point = make_pair(i, j);
                //is this point earth that never been searched?
                if(used.find(point) == used.end() && grid[i][j] == '1'){
                    Q.push(point);
                    used.insert(point);
                    ++ans;
                    //BFS for all earch near it 
                    while(Q.size() > 0){
                        auto node = Q.front();
                        Q.pop();
                        for(int k = 0; k < 4; ++k){
                            int xx = node.first + x_[k];
                            int yy = node.second + y_[k];
                            auto newnode = make_pair(xx, yy);
                            //if new coordinate is valid and is earth and not searched
                            if(xx >= 0 && xx < row 
                               && yy >= 0 && yy < col 
                               && grid[xx][yy] == '1' 
                               && used.find(newnode) == used.end()){
                                Q.push(newnode);
                                used.insert(newnode);
                            }
                        } 
                    }
                }
            }
        }
        return ans;
    }
};